### PR TITLE
Allow provisioning.properties to reference resolvable config

### PR DESCRIPTION
**Pull request by grkvlt**
Enables constructions like the following:
```YAML
- type: vanilla-software-process
  brooklyn.config:
    provisioning.properties:
      minRam: $brooklyn:scopeRoot().config("entity.min.ram")
```

### CODE REVIEW

**Comment by aledsage** [comment]
@grkvlt test failures look plausibly related: failures in `JcloudsLocationTest`.

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java [hunk 1]
```diff
@@ -527,7 +531,7 @@ public ComputeService getComputeService(Map<?,?> flags) {
 
     public ComputeService getComputeService(ConfigBag config) {
         ComputeServiceRegistry registry = getConfig(COMPUTE_SERVICE_REGISTRY);
-        return registry.findComputeService(ResolvingConfigBag.newInstanceExtending(getManagementContext(), config), true);
```
Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: Why are you removing the `ResolvingConfigBag`?

Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: This was added to fix "BROOKLYN-249: JcloudsLocation resolves external config" (see commit 99dd3ba4fc507988bdecd4e57c2b36cbb465b2a5)

**Comment by aledsage** [comment]
The jenkins test failures look plausibly related: `JcloudsLocationTest` tests are failing.

@grkvlt please add test(s) for this. See the existing tests like those in `JcloudsLocationExternalConfigYamlTest`. Can you also confirm that those live tests work (they don't create any VMs, but they go through talking to the cloud for the list of images etc to create the `Template`).

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java (hunk 1 above)
Comment (grkvlt) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: By the time `getComputeService()` is called, I will already have set up the `ConfigBag` to be a resolving config, see `private ConfigBag getConfigBag(Map<?, ?> flags)` method below. The addition of a new `ResolvingConfigBag` was redundant. In essence I'm solving the same problem as **BROOKLYN-249** and also adding the ability to resolve config through the calling entity context, if it exists.

**Comment by grkvlt** [comment]
@aledsage I've fixed the failing tests (there may not be a caller context, so do not rely on its existence) and will add some more to illustrate the use of resolvable entity config in a jclouds location context.

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java [hunk 2]
```diff
@@ -49,6 +50,68 @@
 import javax.annotation.Nullable;
 import javax.xml.ws.WebServiceException;
 
+import org.apache.commons.lang3.ArrayUtils;
```
Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: Can we avoid re-ordering all the imports?

**Comment by aledsage** [comment]
@grkvlt you said that you'd "add some more [tests] to illustrate the use of resolvable entity config in a jclouds location context". Is that something you plan to do for this PR? Do you know when you'll have time to do that?

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java (hunk 2 above)
Comment (grkvlt) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: But this way is more logically defensible!

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java [hunk 3]
```diff
@@ -603,21 +607,37 @@ public MachineLocation obtain(Map<?,?> flags, TemplateBuilder tb) throws NoMachi
         return obtain(MutableMap.builder().putAll(flags).put(TEMPLATE_BUILDER, tb).build());
     }
 
-    /** core method for obtaining a VM using jclouds;
+    private ConfigBag getConfigBag(Map<?, ?> flags) {
```
Comment (ahgittin) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: call this `getConfigBagExtendedWithFlagsResolved` ?

**Comment by ahgittin** [comment]
I think this resolves and stores config too early.  We need to ensure the raw unresolved values are given to the `JcloudsSshMachineLocation` (and a few other places).  See https://github.com/apache/brooklyn-server/pull/194.

Can you check whether `ExternalConfigYamlTest` passes?  They are integration tests (probably shouldn't be!) but I think would illustrate why we can't just resolve everything.

The use case you introduce @grkvlt is useful.  I think we should handle this either by ensuring we resolve each access eg in `buildTemplate` or by maintaining two maps in the code path, `newItemConfigResolved` and `newItemConfigRaw` (instead of `setup`).

Also agree we want a test for this and please undo the huge mad import order change (under what ordering is the new one more sensible??).

Note that the semantics flip once a location (or any adjunct or entity) is managed, `BrooklynObject.config().get()` returns resolved values.  A more elegant but much bigger fix might be to create the `JcloudsSshMachineLocation` early, pass it unresolved config, and then for it to call its own config at which point resolving on access would be a sensible default.  This makes it more like entities (which tend to create the thing they represent, as opposed to the parent creating the thing first which is what `JcloudsLocation` is doing).

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java (hunk 2 above)
Comment (ahgittin) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: @grkvlt just came across another import rearrange you did in 33b0b1c879a9ed07bcdc3138893cd8017e211cfa

can you configure your import order as per
http://brooklyn.apache.org/v/latest/dev/env/ide/index.html#intellij-idea ?

Comment (grkvlt) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: I have done this for all my Brooklyn projects now, thanks for the pointer, @ahgittin

Comment (grkvlt) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: BTW, my ordering was to keep the `org.apache.brooklyn` and `io.brooklyn` and `brooklyn` imports at the bottom, with more generic imports above. But, whatever, am happy to use the standard, have changed my IDE config.

**Comment by aledsage** [comment]
See https://issues.apache.org/jira/browse/BROOKLYN-349 for a description of the underlying problem.

Also see https://github.com/apache/brooklyn-server/pull/341 for an alternative fix and a test for that fix. In PR #341 the resolution is deferred until the value is actually requested. This was a bit more fiddly to implement, but more because it highlighted a problem where the `DslComponent`'s task lost the context of the entity that called it.

We'll only want one of this PR or #341 to be merged, I think. @grkvlt @ahgittin @neykov I'm keen to hear your thoughts on it.

**Comment by aledsage** [comment]
@grkvlt https://github.com/apache/brooklyn-server/pull/341 is now merged, so this PR is no longer needed. Can you close it please (assuming you agree)?

**Pull request closed without merging**